src/website/shared/listeners/cache_suggestions.py

```diff
@@ -295,3 +301,32 @@ def parse_drv_name(name: str) -> tuple[str, str]:
         return match.group(1), match.group(2)
     else:
         return name, ""
+
+
+def maintainers_list(packages: dict, edits: list[MaintainersEdit]) -> list[dict]:
+    """
+    Returns a deduplicated list (by GitHub ID) of all the maintainers of all the
+    affected packages linked to this suggestion, modified by potential
+    user-supplied edits.
+    """
+
+    to_remove = {
+        m.maintainer for m in edits if m.type == MaintainersEdit.EditType.REMOVE
+    }
+    to_add = [m.maintainer for m in edits if m.type == MaintainersEdit.EditType.ADD]
+
+    seen_ids = set()
+    maintainers = list()
+    all_maintainers = [m for pkg in packages.values() for m in pkg["maintainers"]]
+
+    for m in all_maintainers:
+        if m["github_id"] not in seen_ids and m["github_id"] not in to_remove:
+            seen_ids.add(m["github_id"])
+            maintainers.append(m)
+
+    for m in to_add:
+        if m.github_id not in seen_ids and m.github_id not in to_remove:
+            seen_ids.add(m.github_id)
+            maintainers.append(m)
+
+    return maintainers
```

Comment:
Why not... (not tested)

```suggestion
    maintainers = {m for pkg in packages.values() for m in pkg["maintainers"]}
    return list(maintainers - to_remove | to_add)
```

Reply:
I've reworked this part to make it simpler, and relatively idiomatic (and actually fix issues with `dict` vs `non-dict` that were undetected because addition/deletion is just unimplemented for now). There is a comprehension version but I don't like it honestly, it hides a side-effect in an `if`(relying on the fact that `add` returns `None` and this the condition is always `False`):

```
deduped_objects = [obj for obj in objects if obj['id'] not in seen_ids and not seen_ids.add(obj['id'])]
```

I think the current version is reasonably light, but let me know what you think: https://github.com/Nix-Security-WG/nix-security-tracker/pull/553/commits/5b4dd63c2fa73d1879cd09b95ec06175a243b747

Reply:
I think one can easily identify values in the set by adding to the `NixMaintainer` model

```python
    def __hash__(self):
        return hash(self.github_id)
        
    def __eq__(self, other):
        if not isinstance(other, Maintainer):
            return False
        return self.github_id == other.github_id
```

Reply:
The thing is, at this point, we handle dictionaries, that have been generated by the other parts of suggestion caching (the list of maintainers to edit), so those are dicts and not `NixMaintainer`s. IMHO it's not really worth it to bend the rest of the code just to get a one-liner here, as the function might not be the most elegant, but it's relatively short, simple and understandable.

That being said, maybe we want to add those functions anyway, orthogonally; this probably makes sense as a `hash` and `eq` implementation.

Reply:
(as a small aside, it's also less magic - with the custom hash instance, a reader must known that `__hash__` and `__eq__` have be redefined to understand why the code works as intended)